src/cpp/database/database_queue.hpp

```diff
@@ -355,20 +356,48 @@ class DatabaseEntityQueue : public DatabaseQueue<EntityDiscoveryInfo>
     {
         try
         {
-            auto id = database_->insert(front().second.entity);
-            if (EntityKind::HOST  == front().second.entity->kind ||
-                    EntityKind::USER == front().second.entity->kind ||
-                    EntityKind::PROCESS == front().second.entity->kind ||
-                    EntityKind::LOCATOR == front().second.entity->kind)
+            EntityDiscoveryInfo info = front().second;
+
+            // Physical entities
+            if (EntityKind::HOST  == info.entity->kind ||
+                    EntityKind::USER == info.entity->kind ||
+                    EntityKind::PROCESS == info.entity->kind ||
+                    EntityKind::LOCATOR == info.entity->kind)
             {
-                details::StatisticsBackendData::get_instance()->on_physical_entity_discovery(
-                    front().second.domain_id, id,
-                    front().second.entity->kind);
+                info.entity->id = database_->insert(info.entity);
+
+                details::StatisticsBackendData::get_instance()->on_physical_entity_discovery(info.domain_id,
+                        info.entity->id,
+                        info.entity->kind, info.discovery_status);
             }
+            // Domain entities
             else
             {
-                details::StatisticsBackendData::get_instance()->on_domain_entity_discovery(front().second.domain_id, id,
-                        front().second.entity->kind, details::StatisticsBackendData::DISCOVERY);
+                // The topic is never updated/undiscovered, is only discovered. So the status is not changed.
+                // It status will only be updated if its endpoints are discovered/undiscovered.
+                if (info.entity->kind == EntityKind::TOPIC)
+                {
+                    info.entity->id = database_->insert(info.entity);
```

Comment:
We could `assert` that `info.discovery_status == DISCOVERY` and that `info.entity->id.is_valid_and_unique()`